clustermesh-apiserver/mcsapi-coredns-cfg/root.go

```diff
@@ -0,0 +1,172 @@
+			err = fmt.Errorf("Corefile not found in ConfigMap %s/%s", config.CoreDNSNamespace, config.CoreDNSConfigMapName)
+			logger.Error(err.Error())
+			return
+		}
+
+		if corefile, err = updateCorefile(config.CoreDNSClusterDomain, config.CoreDNSClustersetDomain, corefile); err != nil {
+			logger.Error("Failed to update CoreDNS Corefile", logfields.Error, err)
+			return
+		}
+		if corefile == "" {
+			logger.Info("CoreDNS might already have MCS-API configuration, skipping configuration")
+			return
+		}
+		coreDNSConfigmap.Data["Corefile"] = corefile
+
+		if _, err = client.CoreV1().ConfigMaps(config.CoreDNSNamespace).Update(ctx, coreDNSConfigmap, metav1.UpdateOptions{}); err != nil {
+			logger.Error("Failed to update CoreDNS ConfigMap", logfields.Error, err)
+			return
+		}
+
+		logger.Info("CoreDNS ConfigMap was updated successfully")
+
+		err = restartCoreDNS(ctx, client, config)
+		if err != nil {
+			logger.Error("Failed to restart CoreDNS Deployment", logfields.Error, err)
+		} else {
+			logger.Info("CoreDNS is rolling out with the new configuration")
+		}
+		return
+	}))
+}
+
+func updateCorefile(clusterDomain, clustersetDomain string, corefile string) (string, error) {
+	if strings.Contains(corefile, clustersetDomain) || strings.Contains(corefile, "multicluster") {
+		return "", nil // This is not an error as this command might have already been executed
+	}
+
+	clusterDomainEscaped := strings.ReplaceAll(clusterDomain, ".", "\\.")
+	kubernetesMatchRegex, err := regexp.Compile(fmt.Sprintf(`(?m)^\s*kubernetes.*%s.*\{`, clusterDomainEscaped))
+	if err != nil {
+		return "", fmt.Errorf("Failed to compile regex for kubernetes plugin matching: %w", err)
+	}
+	if !kubernetesMatchRegex.MatchString(corefile) {
+		return "", fmt.Errorf("CoreDNS not configured with kubernetes plugin and the domain '%s'", clusterDomain)
+	}
+
+	corefile = strings.ReplaceAll(
+		corefile,
+		clusterDomain,
+		clusterDomain+" "+clustersetDomain,
+	)
+	kubernetesReplaceRegex := regexp.MustCompile(`(?m)^(\s*)kubernetes(.*)\{`)
+	corefile = kubernetesReplaceRegex.ReplaceAllString(
+		corefile,
+		fmt.Sprintf("${1}kubernetes${2}{\n${1}   multicluster %s", clustersetDomain),
+	)
+	return corefile, nil
+}
+
+func restartCoreDNS(ctx context.Context, client client.Clientset, config coreDNSConfig) error {
+	existingDeployment, err := client.AppsV1().Deployments(config.CoreDNSNamespace).Get(ctx, config.CoreDNSDeploymentName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf(
+			"Failed to get CoreDNS Deployment %s/%s: %w",
+			config.CoreDNSNamespace, config.CoreDNSDeploymentName, err,
+		)
+	}
+
+	if existingDeployment.Spec.Paused {
+		return fmt.Errorf(
+			"Deployment %s/%s is paused",
+			config.CoreDNSNamespace, config.CoreDNSDeploymentName,
+		)
+	}
+
+	_, err = client.AppsV1().Deployments(config.CoreDNSNamespace).Apply(
+		ctx,
+		applyappsv1.Deployment(config.CoreDNSDeploymentName, config.CoreDNSNamespace).
+			WithAnnotations(map[string]string{
```

Comment:
Just realized that I had forgotten a piece in my previous suggestion:

```suggestion
		applyappsv1.Deployment(config.CoreDNSDeploymentName, config.CoreDNSNamespace).
			WithSpec(applyappsv1.DeploymentSpec().WithTemplate(
				v1.PodTemplateSpec().WithAnnotations(map[string]string{
```

Reply:
Ah yep indeed :sweat_smile: